public/apps/configuration/utils/list-state-utils.tsx

```diff
@@ -0,0 +1,128 @@
+
+/**
+ * Update an element in an sub array
+ * @param setStateCallback setState function
+ * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern" 
+ * @param newValue value to be updated
+ *
+ * e.g. 
+ *  currentState = [{a: 1}, {a: 2}]
+ *  path = [1, 'a']
+ *  newValue = 10
+ * The newState will be [{a: 1}, {a: 10}]
+ */
+export function updateElementInArray<T>(
+  setStateCallback: Dispatch<SetStateAction<any[]>>,
+  path: StringRepresentable|StringRepresentable[],
+  newValue: T | (() => T)
+) {
+  setStateCallback(prevState => {
+    let newState = [...prevState];
+    set(newState, path, resolveValue(newValue));
+    return newState;
+  });
+}
+
+export const updateElementInArrayHandler = curry(updateElementInArray);
+
+/**
+ * Append an element to an array (or sub array indicating by path)
+ * @param setStateCallback setState function
+ * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern", use [] to indicate root level
+ * @param newValue value to be updated
+ *
+ * e.g.
+ * Scenario 1, path = [] to append to root level array
+ *  currentState = [1, 2]
+ *  path = []
+ *  newValue = 3
+ * The newState will be [1, 2, 3]
+ *
+ * Scenario 2, path != [] to append to sub array
+ *  currentState = [{a: [1, 2]}, {a: [3, 4]}]
+ *  path = [0, 'a']
+ *  newValue = 5
+ * The newState will be [{a: [1, 2, 5]}, {a: [3, 4]}]
+ */
+export function appendElementToArray<T>(
+  setStateCallback: Dispatch<SetStateAction<any[]>>,
+  path: StringRepresentable|StringRepresentable[],
+  newValue: T | (() => T)
+) {
+  const resolvedNewValue = resolveValue(newValue);
+  setStateCallback(prevState => {
+    if ((path as StringRepresentable[]).length == 0) {
+      let newState = [...prevState, resolvedNewValue];
+      return newState;
+    }
+    else {
+      const newArray = [...(get(prevState, path) as T[]), resolvedNewValue];
+      let newState = [...prevState];
+      set(newState, path, newArray);
+      return newState;
+    }
+  });
+}
+
+/**
+ * Remove an element from an array (or sub array indicating by path)
+ * @param setStateCallback setState function
+ * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern", use [] to indicate root level
+ * @param index index of element to be removed
+ *
+ * e.g.
+ * Scenario 1, path = [] to append to root level array
+ *  currentState = [1, 2]
+ *  path = []
+ *  index = 0
+ * The newState will be [2]
+ *
+ * Scenario 2, path != [] to append to sub array
+ *  currentState = [{a: [1, 2]}, {a: [3, 4]}]
+ *  path = [0, 'a']
+ *  index = 1
+ * The newState will be [{a: [1]}, {a: [3, 4]}]
+ */
+export function removeElementFromArray<T>(
+  setStateCallback: Dispatch<SetStateAction<any[]>>,
+  path: StringRepresentable|StringRepresentable[],
+  index: number
+) {
+  setStateCallback(prevState => {
+    if ((path as StringRepresentable[]).length == 0) {
+      let newState = [...prevState];
+      newState.splice(index, 1);
+      return newState;
+    }
+    else {
```

Comment:
could prettier help here to format code to 
```} else {```

Reply:
Improved.